Feature/profile search

ilililililililililililililil opened the pull request:
If commits will apply it'll add new EndPoint /api/v1/profiles/ 
for profiles search by location, sex, age

request example:
/api/v1/profiles/?telegram_id=187392253&age_lt=40&age_gte=28&sex=Парень&location=Москва

request should be sent using telegram_id which was registered as user (exists in db), 
params age_lt and age_gte should be defined
same about locations - it should be imported from fixtures,
sex and location should be mapped with constants defined in backend area:

class Sex(models.TextChoices):
    """
    Константы (пол пользователя).
    """

    MAN = ("Парень", "Парень")
    WOMAN = ("Девушка", "Девушка")

Commit:
Add new EP /api/v1/profiles/ for profiles search
by location, sex, age

Commit:
Add new EP /api/v1/profiles/ for profiles search
by location, sex, age

ilililililililililililililil committed:
Merge branch 'develop' into feature/profile-search

ArtemKAF commented on src/backend/search/views.py:
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/83e79c4fa2d22f7420f866235854e70a53ae254e/src/backend/search/views.py#L55-L74
А зачем столько кода писать? Есть готовые дженерики или миксины, в которых это уже работает и нужно только парой строк кода настроить под нюансы проекта.

ilililililililililililililil replied:
> https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/83e79c4fa2d22f7420f866235854e70a53ae254e/src/backend/search/views.py#L55-L74
> 
> 
> А зачем столько кода писать? Есть готовые дженерики или миксины, в которых это уже работает и нужно только парой строк кода настроить под нюансы проекта.


У меня не "столько кода", у меня всего 5 строк начиная с age_lt, так как предыдущие строки нужны для проверки, что запрос пришел от зарегистрированного юзера (даже эту проверку я могу написать в 1 строку это не принципиально).

Итого мы имеем "столько кода" т.е. 5 строк, против "пары строк"(кстати под парой понимается ровно 2 или 2-3 строки ?), в итоге мы имеем превышение всего на 2-3 строки кода, от "идеального варианта".

Мое решение протестировано и работает, кстати и так на дженерике ListAPIView. 

Напиши более конкретно, где можно улучшить и как(не общими словами, а конкретно, если миксины, то какие, если дженерики, то какие (ModelViewSet? или ListModelMixin? )), я посмотрю и возможно скорректирую или мы обсудим.

ArtemKAF replied:
Сори, из-за невнимательности некорректно поставил вопрос. Зачем в готовом методе дженерика generics.ListAPIView переопределять его базовый метод, чтобы реализовать фильтрацию, если есть готовые решения, как фильтрсеты, которые именно для этой задачи создавались?
К примеру:
https://django-filter.readthedocs.io/en/stable/ref/filterset.html
Эта библиотека уже присутствует в проекте.

ilililililililililililililil replied:
> Сори, из-за невнимательности некорректно поставил вопрос. Зачем в готовом методе дженерика generics.ListAPIView переопределять его базовый метод, чтобы реализовать фильтрацию, если есть готовые решения, как фильтрсеты, которые именно для этой задачи создавались? К примеру: https://django-filter.readthedocs.io/en/stable/ref/filterset.html Эта библиотека уже присутствует в проекте.

скорректировал, посмотри плз

ArtemKAF replied:
Хорошо. Что на счет того, чтобы фиксировать просмотренные профиля?

TomatoInOil replied:
> Хорошо. Что на счет того, чтобы фиксировать просмотренные профиля?

Я думаю, что это лучше осуществлять после выполнения действия: лайк, дизлайк

UPD: А ещё хотелось бы, чтобы эндпоинт был идемпотентным

ilililililililililililililil replied:
> UPD: А ещё хотелось бы, чтобы эндпоинт был идемпотентным

а сейчас он не такой ?

TomatoInOil replied:
> > UPD: А ещё хотелось бы, чтобы эндпоинт был идемпотентным
> 
> а сейчас он не такой ?

Такой.

Commit:
Add fix for new EP /api/v1/profiles/ for profiles search
by location, sex, age

ArtemKAF commented on src/backend/search/filters.py:
Фильтрацию по возрасту забыл добавить.

Commit:
Add exclude viewers for new EP /api/v1/profiles/ for profiles search
by location, sex, age.

Commit:
Add age filter for new EP /api/v1/profiles/ for profiles search
by location, sex, age.

TomatoInOil commented on src/backend/search/views.py:
```diff
@@ -36,7 +43,24 @@ def get_queryset(self):
+        user = UserFromTelegram.objects.get(
+            telegram_id=self.request.query_params.get("telegram_id", None))
+        if user is None:
+            raise exceptions.NotFound("Такого пользователя не существует.")
```
Если не указать `telegram_id` в _query_-параметрах, то падает с исключением `DoesNotExist`. Можно перехватить исключение https://docs.djangoproject.com/en/5.0/ref/exceptions/#objectdoesnotexist и обработать, как ты уже написал
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/210ba8dd7608bf4ebae3d8e577f4ff5131ec47f4/src/backend/search/views.py#L61

TomatoInOil commented on src/backend/search/views.py:
```diff
@@ -36,7 +43,24 @@ def get_queryset(self):
+        if user is None:
+            raise exceptions.NotFound("Такого пользователя не существует.")
+
+        return super().get_queryset().exclude(pk=Profile.objects.all().filter(viewers=user))
```
Здесь используется `field lookup` [exact](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#exact) для pk в методе exclude(), но в аргумент передаётся queryset, который может содержать несколько объектов в выборке, будет вызвано исключение `ValueError`
Чтобы этого избежать можно использовать `field lookup` [in](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#in) .

TomatoInOil commented on src/backend/search/filters.py:
```diff
@@ -0,0 +1,13 @@
+class ProfilesSearchFilterSet(FilterSet):
+    location = ModelChoiceFilter(field_name='location',
+                                 queryset=Location.objects.all())
+    sex = ModelChoiceFilter(field_name='sex')
```
Для `ModelChoiceFilter` аргумент `queryset` обязательный.
https://django-filter.readthedocs.io/en/stable/ref/filters.html#modelchoicefilter-and-modelmultiplechoicefilter-arguments

UPD: Можно заменить на `CharFilter`, для `location` тоже можно использовать `CharFilter`, если в `field_name` передавать `"location__name"`

ilililililililililililililil replied:
В итоге для location я решил оставить ModelChoiceFilter,
а для sex заменил на ChoiceFilter
вот в таком виде (sex = ChoiceFilter(field_name='sex', choices=Sex.choices))
https://django-filter.readthedocs.io/en/stable/ref/filters.html#choicefilter

TomatoInOil commented on src/backend/search/views.py:
Если посмотреть в логи при совершении запроса (в ответе 3 объекта), то можно увидеть следующее
```
<...some queries...>
2024-02-15 16:50:01,155 - django.db.backends - DEBUG - (0.000) SELECT "profiles_userfromtelegram"."id", "profiles_userfromtelegram"."telegram_id", "profiles_userfromtelegram"."residence_id" FROM "profiles_userfromtelegram" WHERE "profiles_userfromtelegram"."id" = 1 LIMIT 21; args=(1,); alias=default
<...some queries...>
2024-02-15 16:50:01,157 - django.db.backends - DEBUG - (0.000) SELECT "profiles_userfromtelegram"."id", "profiles_userfromtelegram"."telegram_id", "profiles_userfromtelegram"."residence_id" FROM "profiles_userfromtelegram" WHERE "profiles_userfromtelegram"."id" = 3 LIMIT 21; args=(3,); alias=default
<...some queries...>
2024-02-15 16:50:01,159 - django.db.backends - DEBUG - (0.000) SELECT "profiles_userfromtelegram"."id", "profiles_userfromtelegram"."telegram_id", "profiles_userfromtelegram"."residence_id" FROM "profiles_userfromtelegram" WHERE "profiles_userfromtelegram"."id" = 2 LIMIT 21; args=(2,); alias=default
<...some queries...>
```
Для каждого профиля совершается запрос на получение полей `telegram_id` и `residence_id` из таблицы `profiles_userfromtelegram`. Можно избежать такого повторения, используя [select_related("user")](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#select-related)
Тоже самое можно сказать про `location.name`, поэтому `select_related("user", "location")`

TomatoInOil commented on src/backend/search/views.py:
В ответе должны отсутствовать профили, которые скрыты из поиска, то есть `is_visible=False`, сейчас `queryset` не фильтруется по этому полю.

TomatoInOil commented on src/backend/search/views.py:
```diff
@@ -36,7 +43,24 @@ def get_queryset(self):
+    queryset = Profile.objects.all()
+    serializer_class = ProfileSerializer
+    filterset_class = ProfilesSearchFilterSet
+    http_method_names = ["get"]
+
```
Даже без этого переопределения будут доступны только методы `GET, HEAD, OPTIONS`, потому что разрешенные методы определяются динамично в зависимости от наличия в классе метода, который обрабатывает этот тип запроса. 
https://github.com/django/django/blob/22285d366c9061a668003638613685f5a135a4c3/django/views/generic/base.py#L178-L179

TomatoInOil reviewed (changes requested):
Хорошо было бы покрыть ручку тестами.

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add select related
- add is_visible filter
- add fix for exclude
- add fix for object DoesNotExist

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add select related
- add is_visible filter
- add fix for exclude
- add fix for object DoesNotExist

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add select related
- add is_visible filter
- add fix for exclude
- add fix for object DoesNotExist

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add select related
- add is_visible filter
- add fix for exclude
- add fix for object DoesNotExist

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add ChoiceFilter for ProfilesSearchFilterSet

ilililililililililililililil commented:
> Хорошо было бы покрыть ручку тестами.

готов сделать в качестве второй задачи

Commit:
Add several fixes for for new EP /api/v1/profiles/ for profiles search
by location, sex, age.
- add ChoiceFilter for ProfilesSearchFilterSet
